src/native/mqtt_connection.c

```diff
@@ -245,6 +243,65 @@ static void s_on_connection_interrupted(
     /********** JNI ENV RELEASE **********/
 }
 
+static void s_on_connection_success(
+    struct aws_mqtt_client_connection *client_connection,
+    enum aws_mqtt_connect_return_code return_code,
+    bool session_present,
+    void *user_data) {
+    (void)client_connection;
+    (void)return_code;
+
+    struct mqtt_jni_connection *connection = user_data;
+
+    /********** JNI ENV ACQUIRE **********/
+    JNIEnv *env = aws_jni_acquire_thread_env(connection->jvm);
+    if (env == NULL) {
+        /* If we can't get an environment, then the JVM is probably shutting down.  Don't crash. */
+        return;
+    }
+    jobject mqtt_connection = (*env)->NewLocalRef(env, connection->java_mqtt_connection);
+    if (mqtt_connection) {
+
+        (*env)->CallVoidMethod(env, mqtt_connection, mqtt_connection_properties.on_connection_success, session_present);
+
+        (*env)->DeleteLocalRef(env, mqtt_connection);
+
+        AWS_FATAL_ASSERT(!aws_jni_check_and_clear_exception(env));
+    }
+    aws_jni_release_thread_env(connection->jvm, env);
+    /********** JNI ENV RELEASE **********/
+
+    s_mqtt_jni_connection_release(connection);
+}
+
+static void s_on_connection_failure(
+    struct aws_mqtt_client_connection *client_connection,
+    int error_code,
+    void *user_data) {
+    (void)client_connection;
+
+    struct mqtt_jni_connection *connection = user_data;
+
+    /********** JNI ENV ACQUIRE **********/
+    JNIEnv *env = aws_jni_acquire_thread_env(connection->jvm);
+    if (env == NULL) {
+        /* If we can't get an environment, then the JVM is probably shutting down.  Don't crash. */
+        return;
+    }
+    jobject mqtt_connection = (*env)->NewLocalRef(env, connection->java_mqtt_connection);
+    if (mqtt_connection) {
+        (*env)->CallVoidMethod(env, mqtt_connection, mqtt_connection_properties.on_connection_failure, error_code);
+
+        (*env)->DeleteLocalRef(env, mqtt_connection);
+
+        AWS_FATAL_ASSERT(!aws_jni_check_and_clear_exception(env));
+    }
+    aws_jni_release_thread_env(connection->jvm, env);
+    /********** JNI ENV RELEASE **********/
+
+    s_mqtt_jni_connection_release(connection);
```

Comment:
This should be deleted as well.

These callbacks should not be mutating references in any way.